File: src/lib/helix/router.ts
import { cosine, embed } from "./embeddings";
import type { Intent } from "./types";

const PROTOTYPES: { intent: Intent; text: string }[] = [
  { intent: "regime", text: "regime trend bull bear high vol range sma market state" },
  { intent: "anomaly", text: "anomaly dump spike inflow funding zscore unusual stress" },
  { intent: "backtest", text: "backtest sma crossover strategy sharpe drawdown trades" },
  { intent: "risk", text: "risk var cvar shock scenario pnl portfolio drop crash" },
  { intent: "research", text: "cite evidence research note on-chain why explain grounded" },
  { intent: "portfolio", text: "portfolio review book weights allocation ballast copilot" },
  { intent: "eval", text: "eval metrics recall ndcg promotion gates golden regression" },
  { intent: "training", text: "train lora qlora checkpoint dataset sft reranker registry" },
  { intent: "memory", text: "remember forget recall my mandate risk cap note fact" },
  { intent: "general", text: "hello help what can you do helix voice agent mcp tools" },
];

const VEC = PROTOTYPES.map((p) => ({ intent: p.intent, v: embed(p.text) }));

export function classifyIntent(query: string): Intent {
  const q = query.toLowerCase();
  if (/\b(backtest|sharpe|sma crossover|cagr)\b/.test(q)) return "backtest";
  if (/\b(var|cvar|shock|drop 12|risk if|scenario)\b/.test(q)) return "risk";
  if (/\b(anomal|inflow|dump|funding|z-score|zscore|stress window)\b/.test(q)) return "anomaly";
  if (/\b(regime|bull|bear|high-vol|high vol)\b/.test(q)) return "regime";
  if (/\b(promote|ndcg|recall@|golden|evalforge|eval)\b/.test(q)) return "eval";
  if (/\b(lora|qlora|checkpoint|train|sft|registry)\b/.test(q)) return "training";
  if (/\b(remember|forget|what do you remember)\b/.test(q)) return "memory";
  if (/\b(portfolio|book|weights|copilot review)\b/.test(q)) return "portfolio";
  if (/\b(cite|evidence|why|research|staking|on-chain|onchain)\b/.test(q)) return "research";
  const qv = embed(query);
  let best: Intent = "general";
  let score = -1;
  for (const p of VEC) {
    const s = cosine(qv, p.v);
    if (s > score) {
      score = s;
      best = p.intent;
    }
  }
  return score < 0.12 ? "general" : best;
}

File: python/helix/router.py
"""Supervisor intent classifier: lexical rules first, then embedding prototypes."""

from __future__ import annotations

import re

from helix.embeddings import cosine, embed
from helix.models import Intent

PROTOTYPES: list[tuple[Intent, str]] = [
    ("regime", "regime trend bull bear high vol range sma market state"),
    ("anomaly", "anomaly dump spike inflow funding zscore unusual stress"),
    ("backtest", "backtest sma crossover strategy sharpe drawdown trades"),
    ("risk", "risk var cvar shock scenario pnl portfolio drop crash"),
    ("research", "cite evidence research note on-chain why explain grounded"),
    ("portfolio", "portfolio review book weights allocation ballast copilot"),
    ("eval", "eval metrics recall ndcg promotion gates golden regression"),
    ("training", "train lora qlora checkpoint dataset sft reranker registry"),
    ("general", "hello help what can you do helix voice agent mcp tools"),
]
_VEC = [(intent, embed(text)) for intent, text in PROTOTYPES]

_RULES: list[tuple[Intent, re.Pattern[str]]] = [
    ("backtest", re.compile(r"\b(backtest|sharpe|sma crossover|cagr)\b", re.I)),
    ("risk", re.compile(r"\b(var|cvar|shock|drop 12|risk if|scenario|preflight|kill.?switch)\b", re.I)),
    ("anomaly", re.compile(r"\b(anomal|inflow|dump|funding|z-score|zscore|stress window|open interest)\b", re.I)),
    ("regime", re.compile(r"\b(regime|bull|bear|high-vol|high vol|klines|mark price)\b", re.I)),
    ("eval", re.compile(r"\b(promote|ndcg|recall@|golden|evalforge|eval)\b", re.I)),
    ("training", re.compile(r"\b(lora|qlora|checkpoint|train|sft|registry)\b", re.I)),
    ("portfolio", re.compile(r"\b(portfolio|book|weights|copilot review|position|balance|account|futures)\b", re.I)),
    ("research", re.compile(r"\b(cite|evidence|why|research|staking|on-chain|onchain)\b", re.I)),
]


def classify_intent(query: str) -> Intent:
    for intent, pat in _RULES:
        if pat.search(query):
            return intent
    qv = embed(query)
    best: Intent = "general"
    score = -1.0
    for intent, vec in _VEC:
        s = cosine(qv, vec)
        if s > score:
            score, best = s, intent
    return "general" if score < 0.12 else best
